Foundation/String.swift

```diff
@@ -29,15 +29,20 @@ extension String : _ObjectTypeBridgeable {
             result = source._storage
         } else if type(of: source) == _NSCFString.self {
             let cf = unsafeBitCast(source, to: CFString.self)
-            let str = CFStringGetCStringPtr(cf, CFStringEncoding(kCFStringEncodingUTF8))
-            if str != nil {
-                result = String(cString: str!)
+            let length = CFStringGetLength(cf) // This value is not always the length in characters, in spite of what the documentation says
+
+            // FIXME: If we had a reliable way to determine the length of `cf`
+            //        in bytes, then we wouldn't have to allocate a new buffer
+            //        if `CFStringGetCStringPtr(_:_:)` doesn't return nil.
+            if let buffer = CFStringGetCharactersPtr(cf) {
```

Comment:
The recommended pattern is documented here:

https://developer.apple.com/library/content/documentation/CoreFoundation/Conceptual/CFStrings/Articles/AccessingContents.html

Reply:
Specifically:

```
CFStringRef str;
const UniChar *chars;
 
str = CFStringCreateWithCString(NULL, "Hello World", kCFStringEncodingMacRoman);
chars = CFStringGetCharactersPtr(str);
if (chars == NULL) {
    CFIndex length = CFStringGetLength(str);
    UniChar *buffer = malloc(length * sizeof(UniChar));
    CFStringGetCharacters(str, CFRangeMake(0, length), buffer);
    // Process the characters...
    free(buffer);
}
```

Reply:
As I write above in the description, this recommended pattern cannot be used because `CFStringGetCharacters` is buggy and returns bytes instead of UTF16 characters; attempting to use the recommended pattern clobbers multibyte Unicode strings. We must here check if `CFStringGetCharactersPtr` returns `NULL`, and if so, retrieve UTF8-encoded bytes instead.